### PR TITLE
tests: mark distributions and outputs

### PR DESCRIPTION
**Pull request by behrmann**
This allows for fine-grained choices of which integration tests to run

### CODE REVIEW

File: tests/test_boot.py [hunk 1]
```diff
@@ -10,8 +10,15 @@
...
 @pytest.mark.integration
-@pytest.mark.parametrize("format", [f for f in OutputFormat if f != OutputFormat.none])
-def test_boot(format: OutputFormat) -> None:
+@pytest.mark.parametrize(
+    "format",
+    [pytest.param(f, marks=getattr(pytest.mark, f"output_{f}")) for f in OutputFormat if f != OutputFormat.none],
+)
+@pytest.mark.parametrize(
+    "distribution",
+    [pytest.param(d, marks=getattr(pytest.mark, f"distro_{d}")) for d in Distribution if d != Distribution.custom],
```
Comment (keszybz) [inline tests/test_boot.py:19]: This gives me:
```python
[ParameterSet(values=(<Distribution.fedora: 'fedora'>,), marks=(MarkDecorator(mark=Mark(name='distro_fedora', args=(), kwargs={})),), id=None),
 ParameterSet(values=(<Distribution.debian: 'debian'>,), marks=(MarkDecorator(mark=Mark(name='distro_debian', args=(), kwargs={})),), id=None),
 ParameterSet(values=(<Distribution.ubuntu: 'ubuntu'>,), marks=(MarkDecorator(mark=Mark(name='distro_ubuntu', args=(), kwargs={})),), id=None),
 ParameterSet(values=(<Distribution.arch: 'arch'>,), marks=(MarkDecorator(mark=Mark(name='distro_arch', args=(), kwargs={})),), id=None),
 ParameterSet(values=(<Distribution.opensuse: 'opensuse'>,), marks=(MarkDecorator(mark=Mark(name='distro_opensuse', args=(), kwargs={})),), id=None),
 ParameterSet(values=(<Distribution.mageia: 'mageia'>,), marks=(MarkDecorator(mark=Mark(name='distro_mageia', args=(), kwargs={})),), id=None),
 ParameterSet(values=(<Distribution.centos: 'centos'>,), marks=(MarkDecorator(mark=Mark(name='distro_centos', args=(), kwargs={})),), id=None),
 ParameterSet(values=(<Distribution.rhel: 'rhel'>,), marks=(MarkDecorator(mark=Mark(name='distro_rhel', args=(), kwargs={})),), id=None),
 ParameterSet(values=(<Distribution.rhel_ubi: 'rhel-ubi'>,), marks=(MarkDecorator(mark=Mark(name='distro_rhel-ubi', args=(), kwargs={})),), id=None),
 ParameterSet(values=(<Distribution.openmandriva: 'openmandriva'>,), marks=(MarkDecorator(mark=Mark(name='distro_openmandriva', args=(), kwargs={})),), id=None),
 ParameterSet(values=(<Distribution.rocky: 'rocky'>,), marks=(MarkDecorator(mark=Mark(name='distro_rocky', args=(), kwargs={})),), id=None),
 ParameterSet(values=(<Distribution.alma: 'alma'>,), marks=(MarkDecorator(mark=Mark(name='distro_alma', args=(), kwargs={})),), id=None),
 ParameterSet(values=(<Distribution.gentoo: 'gentoo'>,), marks=(MarkDecorator(mark=Mark(name='distro_gentoo', args=(), kwargs={})),), id=None)]
```
I don't want to know. If it works, that must be good enough ;)

File: pytest.ini [hunk 2]
```diff
@@ -1,4 +1,24 @@
...
+    distro_arch: tests concerning Arch
+    distro_opensuse: tests concerning OpenSUSE
+    distro_mageia: tests concerning Mageia
+    distro_centos: tests concerning CentOS
+    distro_rhel: tests concerning RHEL
+    distro_rhel_ubi: tests concerning RHEL-UBI
+    distro_openmandriva: tests concerning OpenMandriva
+    distro_rocky: tests concerning Rocky Linux
+    distro_alma: tests concerning Alma Linux
+    distro_gentoo: tests concerning Gentoo
```
Comment (keszybz) [inline pytest.ini:16]: The first line has a dot, the others don't. Please drop the dot.

The word "concerning" is concerning me. Can we be more precise here? E.g. "tests building an image with Fedora", and "tests creating a tar output"?

**Review by keszybz — CHANGES_REQUESTED** [review]

File: tests/conftest.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        type=parse_integration_distribution,
+    )
+    parser.addoption(
+        "-O",
+        "--integration-output",
+        metavar="OUTPUT",
+        help="Run the integration tests for the given OUTPUT.",
+        default="all",
+        type=parse_integration_output,
+    )
```
Comment (DaanDeMeyer) [inline tests/conftest.py:35]: I would still drop this, you can already do this via the test name "test_boot[disk]". Until we add more tests parametrized over the output format we shouldn't add this.

File: tests/conftest.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+            return {detect_distribution()[0]}
+        return {Distribution(d) for d in s.split(",")}
+
+    def parse_integration_output(s: str) -> set[OutputFormat]:
+        if s == "all":
+            return {f for f in OutputFormat if f != OutputFormat.none}
+        return {OutputFormat(o) for o in s.split(",")}
+
+    parser.addoption(
+        "-D",
```
Comment (DaanDeMeyer) [inline tests/conftest.py:21]: Why not a lower case d?

File: tests/conftest.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
+from typing import Any, Optional
```
Comment (DaanDeMeyer) [inline tests/conftest.py:1]: Missing SPDX line

File: tests/test_boot.py [hunk 6]
```diff
@@ -12,14 +12,22 @@
 
 
 @pytest.mark.integration
-@pytest.mark.parametrize("format", [f for f in OutputFormat if f != OutputFormat.none])
-def test_boot(format: OutputFormat) -> None:
+@pytest.mark.parametrize(
+    "format",
+    [pytest.param(f, marks=pytest.mark.output(f)) for f in OutputFormat if f != OutputFormat.none],
```
Comment (DaanDeMeyer) [inline tests/test_boot.py:17]: Hmm, we should probably just run the test for "none" as well but just stop after the build step like the "tar" output.

File: tests/conftest.py [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+        return {Distribution(d) for d in s.split(",")}
+
+    def parse_integration_output(s: str) -> set[OutputFormat]:
+        if s == "all":
+            return {f for f in OutputFormat if f != OutputFormat.none}
+        return {OutputFormat(o) for o in s.split(",")}
+
+    parser.addoption(
+        "-D",
+        "--integration-distribution",
```
Comment (DaanDeMeyer) [inline tests/conftest.py:22]: I'd just name this --distribution

File: tests/conftest.py [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+    config.addinivalue_line(
+        "markers", "output(output): mark test as belonging to given output."
+    )
+
+
+def pytest_runtest_setup(item: Any) -> None:
+    requested_distros = item.config.getoption("--integration-distribution")
+    marked_distros = {mark.args[0] for mark in item.iter_markers(name="distribution")}
+    if marked_distros and not marked_distros.intersection(requested_distros):
+        pytest.skip(f"Selected distributions are {requested_distros}")
```
Comment (DaanDeMeyer) [inline tests/conftest.py:51]: I'm a bit confused about this line?

**Review by DaanDeMeyer — CHANGES_REQUESTED** [review]

File: tests/conftest.py (hunk 3 above)
Comment (behrmann) [reply tests/conftest.py]: Dropped it, although even so I found it much more convenient then typing out the test instance name

File: tests/conftest.py (hunk 5 above)
Comment (behrmann) [reply tests/conftest.py:2]: Added.

File: tests/conftest.py [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+from typing import Any, Optional
+
+import pytest
+
+from mkosi.config import OutputFormat
+from mkosi.distributions import Distribution, detect_distribution
+
+
+def pytest_addoption(parser: Any) -> None:
+    def parse_integration_distribution(s: str) -> set[Optional[Distribution]]:
```
Comment (DaanDeMeyer) [inline tests/conftest.py:11]: This can be named `parse_distribution()` now

File: tests/conftest.py (hunk 4 above)
Comment (behrmann) [reply tests/conftest.py:17]: pytest doesn't seem to use all that many short options, but all of the ones it has are lowercase. I thought it safer to steer clear of lowercase options.

File: tests/test_boot.py (hunk 6 above)
Comment (behrmann) [reply tests/test_boot.py:17]: I think I'd leave that to a later PR. I'd like to split `test_boot` apart a bit, but I'd rather get this PR over the line first.

**Pull request merged**